### DISCUSSION TITLE
Using components with cmdroute

### DISCUSSION THREAD

**Discussion by holedaemon** [diamondburned/arikawa#447] [Help]
Hi, I've been using the library for a while now, I appreciate the work that's been put into it. One thing I've always had trouble with is using it for components. When I say components, I specifically mean Discord's application command components, for clarity.

Today I tried adding a component handler with cmdroute:
```go
r := cmdroute.NewRouter()
r.Use(b.useLogger())

r.AddFunc("help", b.cmdHelp)
r.Sub("listen", func(r *cmdroute.Router) {
	r.AddFunc("list", b.cmdListenList)
	r.AddFunc("add", b.cmdListenAdd)
	r.AddFunc("remove", b.cmdListenRemove)
	r.AddComponentFunc("listen-modify-select", b.cmpListenModify)
})

return r
```

but when I try using the command via Discord, the application doesn't respond. I've read the source code for cmdroute's handling, and it seems like using the custom_id to register the component is the correct way, but may I'm wrong? I also wrote a middleware that logs the `InteractionEvent` as it comes through, to see that we're actually getting something, but nothing happens when I trigger the command.

I'm at a loss here, I'm not sure what I could possibly doing wrong. Any insight would be appreciated.

**Comment by diamondburned** [comment]
How are you using `r`? Are you calling `AddInteractionHandler` [like so](https://github.com/diamondburned/arikawa/blob/v3/0-examples/commands/main.go#L50)?

**Reply by holedaemon** [reply]
Yep, I am. Every other slash command I've added to the router works, it's just this component that fails to respond. Let me know if there's more you need from me to diagnose and I will gladly provide.

**Reply by diamondburned** [reply]
Is this described in the interaction descriptions that you send over to Discord? How does it look like?

**Reply by holedaemon** [reply]
Sorry, are you referring to the `CreateCommandData` that one uses to overwrite commands? Is there anything special one needs to send for components? This is what mine looks like, the `modify` subcommand is the component.

```go
{
		Name:                     "listen",
		Description:              "Change what channels the bot actively listens to and learns from.",
		NoDMPermission:           true,
		DefaultMemberPermissions: discord.NewPermissions(discord.PermissionManageGuild),
		Options: discord.CommandOptions{
			discord.NewSubcommandOption(
				"add",
				"Add channels to learn from. Doesn't add the channel's past messages; re-train to do that.",
				channelOptions()...,
			),
			discord.NewSubcommandOption(
				"remove",
				"Stop learning from the selected channels. Doesn't remove the channel's data; re-train to do that.",
				channelOptions()...,
			),
			discord.NewSubcommandOption(
				"list",
				"List the channels the bot is currently listening to.",
			),
			discord.NewSubcommandOption(
				"modify",
				"Add or remove channels via the select menu UI.",
			),
		},
	},
```
